File: 2. Add Two Numbers (medium)/solution.java

class ListNode {

    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    public static void main(String args[]) {
    }
}

class Solution {

    public static void main(String args[]) {
        Solution solution = new Solution();

        ListNode l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
        ListNode l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
        ListNode result = solution.addTwoNumbers(l1, l2);
        // Print the result list
        while (result != null) {
            System.out.print(result.val + " ");
            result = result.next;
        }
    }

    int carry = 0;

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null && l2 == null && carry == 0) {
            return null;
        }

        int l1Value = (l1 != null) ? l1.val : 0;
        int l2Value = (l2 != null) ? l2.val : 0;

        int sum = l1Value + l2Value + carry;
        carry = sum / 10;
        sum = sum % 10;

        ListNode resultNode = new ListNode(sum);
        resultNode.next = addTwoNumbers((l1 != null) ? l1.next : null, (l2 != null) ? l2.next : null);
        return resultNode;
    }
}

File: 2. Add Two Numbers (medium)/solution.dart
//  Definition for singly-linked list.
class ListNode {
  int val;
  ListNode? next;
  ListNode([this.val = 0, this.next]);
}

class Solution {
  ListNode? addTwoNumbers(ListNode? l1, ListNode? l2, [bool surplus = false]) {
    if ((l1 == null) && (l2 == null) && (surplus == false)) {
    } else {
      int val1 = (l1 != null) ? l1.val : 0;
      int val2 = (l2 != null) ? l2.val : 0;
      int sum = val1 + val2 + (surplus ? 1 : 0);
      surplus = (sum >= 10);
      return ListNode(sum % 10, addTwoNumbers(l1?.next, l2?.next, surplus));
    }
    return null;
  }
}

void main() {
  Solution solution = Solution();
  ListNode l1 = ListNode(2, ListNode(4, ListNode(3)));
  ListNode l2 = ListNode(5, ListNode(6, ListNode(4)));
  print(solution.addTwoNumbers(l1, l2));

  Solution solution2 = Solution();
  ListNode l3 = ListNode(0);
  ListNode l4 = ListNode(0);
  print(solution2.addTwoNumbers(l3, l4));

  Solution solution3 = Solution();
  ListNode l5 = ListNode(
    9,
    ListNode(
      9,
      ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9))))),
    ),
  );
  ListNode l6 = ListNode(9, ListNode(9, ListNode(9, ListNode(9))));
  print(solution3.addTwoNumbers(l5, l6));
}
